# Sales Performance Dashboard

## 1. Project Objective

The objective of this project is to analyze retail sales data and identify key trends and business insights related to sales, profit, customers, products, categories, and regions.

The project involves data cleaning and preprocessing using Python and Pandas, data analysis using SQL, and interactive dashboard development using Power BI.

## 2. Data Source

The dataset used in this project is the Superstore sales dataset sourced from Kaggle.

The dataset contains retail transaction information such as orders, customers, products, sales, quantity, discount, profit, category, sub-category, and region.

## 3. Import Libraries

In [1]:
import pandas as pd
import numpy as np

## 4. Load Dataset

In [2]:
df=pd.read_csv("../data/raw/Superstore.csv", encoding="latin1")

### Preview of Dataset

In [3]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


### Dataset Dimensions

In [4]:
df.shape

(9994, 21)

### column names

In [5]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='str')

### data types

In [6]:
df.dtypes

Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code        int64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

### information

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

### missing values

In [8]:
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

### Duplicate Records

In [9]:
df.duplicated().sum()

np.int64(0)

### Statistical Summary

In [10]:
df.describe()

,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


## 7. Data Cleaning

### Standardizing Column Names

In [11]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
    .str.replace("-", "_")
)

df.columns

Index(['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
       'postal_code', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'sales', 'quantity', 'discount', 'profit'],
      dtype='str')

### cleaning text columns

In [12]:
text_columns = df.select_dtypes(include=["str", "object"]).columns

for col in text_columns:
    df[col] = df[col].astype("string").str.strip()

### Converting Date Columns

In [14]:
df["order_date"] = pd.to_datetime(df["order_date"], errors="coerce")
df["ship_date"] = pd.to_datetime(df["ship_date"], errors="coerce")

In [15]:
df[["order_date", "ship_date"]].dtypes

order_date    datetime64[us]
ship_date     datetime64[us]
dtype: object

In [16]:
df.isnull().sum()

row_id           0
order_id         0
order_date       0
ship_date        0
ship_mode        0
customer_id      0
customer_name    0
segment          0
country          0
city             0
state            0
postal_code      0
region           0
product_id       0
category         0
sub_category     0
product_name     0
sales            0
quantity         0
discount         0
profit           0
dtype: int64

### Removing Duplicate Records

In [17]:
df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

### Validating Numerical Columns

In [18]:
numeric_columns = [
    "sales",
    "quantity",
    "discount",
    "profit"
]

df[numeric_columns].describe()

,sales,quantity,discount,profit
count,9994.000000,9994.000000,9994.000000,9994.000000
mean,229.858001,3.789574,0.156203,28.656896
std,623.245101,2.225110,0.206452,234.260108
min,0.444000,1.000000,0.000000,-6599.978000
25%,17.280000,2.000000,0.000000,1.728750
50%,54.490000,3.000000,0.200000,8.666500
75%,209.940000,5.000000,0.200000,29.364000
max,22638.480000,14.000000,0.800000,8399.976000


In [19]:
df["discount"].min(), df["discount"].max()

(np.float64(0.0), np.float64(0.8))

In [20]:
df["quantity"].min(), df["quantity"].max()

(np.int64(1), np.int64(14))

In [21]:
df[["sales", "profit"]].describe()

,sales,profit
count,9994.000000,9994.000000
mean,229.858001,28.656896
std,623.245101,234.260108
min,0.444000,-6599.978000
25%,17.280000,1.728750
50%,54.490000,8.666500
75%,209.940000,29.364000
max,22638.480000,8399.976000


## 8. Feature Engineering

### Calculating Shipping Days

In [23]:
df["shipping_days"] = (
    df["ship_date"] - df["order_date"]
).dt.days

df[["order_date", "ship_date", "shipping_days"]].head()

,order_date,ship_date,shipping_days
0,2016-11-08,2016-11-11,3
1,2016-11-08,2016-11-11,3
2,2016-06-12,2016-06-16,4
3,2015-10-11,2015-10-18,7
4,2015-10-11,2015-10-18,7


In [24]:
df["year"] = df["order_date"].dt.year

In [25]:
df["month"] = df["order_date"].dt.month

In [26]:
df["year_month"] = df["order_date"].dt.to_period("M").astype(str)

In [27]:
df["profit_margin"] = (
    df["profit"] / df["sales"]
) * 100

In [28]:
df["profit_margin"] = df["profit_margin"].replace(
    [np.inf, -np.inf],
    np.nan
)

In [29]:
df.head()

,row_id,order_id,order_date,ship_date,ship_mode,customer_id,customer_name,segment,country,city,...,product_name,sales,quantity,discount,profit,shipping_days,year,month,year_month,profit_margin
0,1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136,3,2016,11,2016-11,16.00
1,2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820,3,2016,11,2016-11,30.00
2,3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714,4,2016,6,2016-06,47.00
3,4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310,7,2015,10,2015-10,-40.00
4,5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164,7,2015,10,2015-10,11.25


In [30]:
df.shape

(9994, 26)

## 9. Exploratory Data Analysis

### Overall Business Performance

In [32]:
total_sales = df["sales"].sum()

total_sales

np.float64(2297200.8603)

In [33]:
total_profit = df["profit"].sum()

total_profit

np.float64(286397.0217)

In [34]:
total_quantity = df["quantity"].sum()

total_quantity

np.int64(37873)

In [35]:
total_orders = df["order_id"].nunique()

total_orders

5009

In [36]:
total_customers = df["customer_id"].nunique()

total_customers

793

In [37]:
profit_margin = (
    total_profit / total_sales
) * 100

profit_margin

np.float64(12.467217240315605)

### Sales and Profit by Category

In [38]:
category_analysis = (
    df.groupby("category")
    .agg(
        sales=("sales", "sum"),
        profit=("profit", "sum"),
        quantity=("quantity", "sum")
    )
    .sort_values("sales", ascending=False)
)

category_analysis

,sales,profit,quantity
category,,,
Technology,836154.0330,145454.9481,6939
Furniture,741999.7953,18451.2728,8028
Office Supplies,719047.0320,122490.8008,22906


### Sales and Profit by Region

In [39]:
region_analysis = (
    df.groupby("region")
    .agg(
        sales=("sales", "sum"),
        profit=("profit", "sum")
    )
    .sort_values("sales", ascending=False)
)

region_analysis

,sales,profit
region,,
West,725457.8245,108418.4489
East,678781.2400,91522.7800
Central,501239.8908,39706.3625
South,391721.9050,46749.4303


### Sub-Category Performance

In [40]:
subcategory_analysis = (
    df.groupby("sub_category")
    .agg(
        sales=("sales", "sum"),
        profit=("profit", "sum")
    )
    .sort_values("sales", ascending=False)
)

subcategory_analysis

,sales,profit
sub_category,,
Phones,330007.0540,44515.7306
Chairs,328449.1030,26590.1663
Storage,223843.6080,21278.8264
Tables,206965.5320,-17725.4811
Binders,203412.7330,30221.7633
Machines,189238.6310,3384.7569
Accessories,167380.3180,41936.6357
Copiers,149528.0300,55617.8249
Bookcases,114879.9963,-3472.5560


### Top 10 Products by Sales

In [41]:
top_products = (
    df.groupby("product_name")["sales"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_products

product_name
Canon imageCLASS 2200 Advanced Copier                                          61599.824
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind    27453.384
Cisco TelePresence System EX90 Videoconferencing Unit                          22638.480
HON 5400 Series Task Chairs for Big and Tall                                   21870.576
GBC DocuBind TL300 Electric Binding System                                     19823.479
GBC Ibimaster 500 Manual ProClick Binding System                               19024.500
Hewlett Packard LaserJet 3310 Copier                                           18839.686
HP Designjet T520 Inkjet Large Format Printer - 24" Color                      18374.895
GBC DocuBind P400 Electric Binding System                                      17965.068
High Speed Automatic Electric Letter Opener                                    17030.312
Name: sales, dtype: float64

### Top 10 Products by profit

In [42]:
top_profit_products = (
    df.groupby("product_name")["profit"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)

top_profit_products

product_name
Canon imageCLASS 2200 Advanced Copier                                          25199.9280
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind     7753.0390
Hewlett Packard LaserJet 3310 Copier                                            6983.8836
Canon PC1060 Personal Laser Copier                                              4570.9347
HP Designjet T520 Inkjet Large Format Printer - 24" Color                       4094.9766
Ativa V4110MDD Micro-Cut Shredder                                               3772.9461
3D Systems Cube Printer, 2nd Generation, Magenta                                3717.9714
Plantronics Savi W720 Multi-Device Wireless Headset System                      3696.2820
Ibico EPK-21 Electric Binding System                                            3345.2823
Zebra ZM400 Thermal Label Printer                                               3343.5360
Name: profit, dtype: float64

### Loss-Making Products

In [43]:
loss_products = (
    df.groupby("product_name")["profit"]
    .sum()
    .sort_values()
)

loss_products.head(10)

product_name
Cubify CubeX 3D Printer Double Head Print                           -8879.9704
Lexmark MX611dhe Monochrome Laser Printer                           -4589.9730
Cubify CubeX 3D Printer Triple Head Print                           -3839.9904
Chromcraft Bull-Nose Wood Oval Conference Tables & Bases            -2876.1156
Bush Advantage Collection Racetrack Conference Table                -1934.3976
GBC DocuBind P400 Electric Binding System                           -1878.1662
Cisco TelePresence System EX90 Videoconferencing Unit               -1811.0784
Martin Yale Chadless Opener Electric Letter Opener                  -1299.1836
Balt Solid Wood Round Tables                                        -1201.0581
BoxOffice By Design Rectangular and Half-Moon Meeting Room Tables   -1148.4375
Name: profit, dtype: float64

### Monthly Sales Trend

In [44]:
monthly_sales = (
    df.groupby("year_month")["sales"]
    .sum()
    .reset_index()
)

monthly_sales.head()

,year_month,sales
0,2014-01,14236.895
1,2014-02,4519.892
2,2014-03,55691.009
3,2014-04,28295.345
4,2014-05,23648.287


### Monthly Profit

In [45]:
monthly_profit = (
    df.groupby("year_month")["profit"]
    .sum()
    .reset_index()
)

monthly_profit.head()

,year_month,profit
0,2014-01,2450.1907
1,2014-02,862.3084
2,2014-03,498.7299
3,2014-04,3488.8352
4,2014-05,2738.7096


### Customer Segment Performance

In [46]:
segment_analysis = (
    df.groupby("segment")
    .agg(
        sales=("sales", "sum"),
        profit=("profit", "sum"),
        orders=("order_id", "nunique")
    )
    .sort_values("sales", ascending=False)
)

segment_analysis

,sales,profit,orders
segment,,,
Consumer,1.161401e+06,134119.2092,2586
Corporate,7.061464e+05,91979.1340,1514
Home Office,4.296531e+05,60298.6785,909


### Discount and Profit Analysis

In [48]:
discount_analysis = (
    df.groupby("discount")
    .agg(
        sales=("sales", "sum"),
        profit=("profit", "sum")
    )
    .sort_index()
)

discount_analysis

,sales,profit
discount,,
0.00,1.087908e+06,320987.6032
0.10,5.436935e+04,9029.1770
0.15,2.755852e+04,1418.9915
0.20,7.645944e+05,90337.3060
0.30,1.032267e+05,-10369.2774
0.32,1.449346e+04,-2391.1377
0.40,1.164178e+05,-23057.0504
0.45,5.484974e+03,-2493.1111
0.50,5.891854e+04,-20506.4281


### Shipping Performance

In [50]:
shipping_analysis = (
    df.groupby("ship_mode")
    .agg(
        orders=("order_id", "nunique"),
        average_shipping_days=("shipping_days", "mean")
    )
)

shipping_analysis

,orders,average_shipping_days
ship_mode,,
First Class,787,2.182705
Same Day,264,0.044199
Second Class,964,3.238046
Standard Class,2994,5.006535


# 10. Business Insights

### 1. Sales Performance by Category
- Technology generated the highest sales at $836,154.03 and the highest profit at

   $145,454.95.

  
- Office Supplies had the highest quantity sold with 22,906 units and generated

  $122,490.80 in profit.

  
- Furniture generated $741,999.80 in sales but only

  $18,451.27 in profit, indicating comparatively lower profitability.

### 2. Regional Profitability
- The West region generated the highest sales at $725,457.82 and the highest profit at

  $108,418.45.

- The East region ranked second in sales with $678,781.24 and generated

  $91,522.78 in profit.
  
- The Central region generated $501,239.89 in sales and

  $39,706.36 in profit.

  
- The South region had the lowest sales at $391,721.91, but generated

  $46,749.43 in profit, which is higher than Central's profit despite having lower sales.

### 3. Customer Segment Performance
- Consumer is the largest customer segment, contributing approximately $1.16M in sales.
  
- Corporate contributes approximately $0.71M, while Home Office contributes approximately

  $0.43M.
  
- Consumer customers therefore represent the largest revenue opportunity.

### 4. Product Performance
- The Canon imageCLASS 2200 Advanced Copier is the top product by sales and also the top product by profit in the analysis.
- A small number of products contribute substantially more revenue than the remaining products, highlighting the importance of identifying and maintaining high-performing products.
- Loss-making products were identified through product-level profitability analysis and should be reviewed for pricing, discounting, and cost-related issues.

### 5. Discount and Profitability
- The discount-versus-profitability analysis shows that higher discount levels are generally associated with weaker profitability.
- Several high-discount observations have low or negative profit.
- Discount policies should therefore be monitored to ensure that increased sales volume does not reduce overall profitability.

### 6. Shipping Performance
- Standard Class accounts for the majority of orders, with approximately 2,994 orders.
- Standard Class also has the highest average shipping time at approximately 5 days.
- Same Day has the shortest shipping time, while First Class and Second Class provide intermediate delivery times.

### 7. Overall Performance
- The business generated approximately $2.30M in total sales and

  $286K in total profit.
  
- Overall profit margin is approximately 12.47%.
- The analysis indicates that profitability is influenced by product selection, regional performance, customer segments, discount levels, and shipping methods.

# 11.Business Recommendations

1. **Focus on high-performing categories**
   - Continue investing in Technology while monitoring the performance of Furniture and Office Supplies.

2. **Improve regional profitability**
   - Investigate the lower profitability of Central and South regions by analyzing product mix, pricing, discounts, and customer demand.

3. **Review loss-making products**
   - Evaluate pricing, discount rates, and costs for products generating negative profit.

4. **Optimize discount strategy**
   - Monitor high-discount transactions and establish discount thresholds that protect profit margins.

5. **Improve shipping efficiency**
   - Review Standard Class shipping performance and investigate opportunities to reduce average delivery time without significantly increasing logistics costs.

6. **Prioritize valuable customer segments**
   - Continue focusing on the Consumer segment while identifying opportunities to increase revenue from Corporate and Home Office customers.

## 12. Export Cleaned Dataset

In [51]:
df.to_csv(
    "../data/cleaned_superstore.csv",
    index=False
)

In [52]:
import os

os.path.exists("../data/cleaned_superstore.csv")

True

In [1]:
import os

print(os.path.abspath("../data/cleaned_superstore.csv"))

C:\Users\kartik\Desktop\Sales-Performance-Dashboard\data\cleaned_superstore.csv
